File: statistics/python/Matrix.py
# -*- coding: utf-8 -*-
# @Author: Jingyuexing
# @Date:   2019-07-11 23:56:22
# @Last Modified by:   Jingyuexing
# @Last Modified time: 2020-01-13 09:11:30
import Vector


class Matrix:
    """docstring for  Matrix
    row 行数
    col 列数
    data 矩阵数据
    """

    def __init__(self, row=0, col=0, data=[]):
        if isinstance(row, (list, tuple)):
            self.row = len(row)
            self.col = len(row[0])
            self.data = row
            self.shape = (self.row, self.col)
        else:
            self.row = row
            self.col = col
            self.data = []
            for x in range(0, self.row):
                self.data.append([])
                for i in range(0, self.col):
                    self.data[x].append([])
            self.shape = (self.row, self.col)

    def indentity(self):
        '''初始化为单位矩阵

        [description]

        Returns:
            [type] -- [description]
        '''
        for i in range(0, self.col):
            for j in range(0, self.row):
                if i == j:
                    self.data[i][j] = 1
                else:
                    self.data[i][j] = 0
        return self

    def product(self, Mb):
        """矩阵相乘

        其他矩阵和其自身相乘

        Arguments:
            Mb {Matrix} -- 相乘的矩阵

        Returns:
            Matrix -- 相乘后的矩阵
        """
        if isinstance(Mb, Matrix) and self.col == Mb.row:
            i = 0
            tempMatrix = Matrix(self.row, Mb.col)
            while(i < self.row):
                j = 0
                while j < Mb.col:
                    tempMatrix.data[i][j] = 0
                    n = 0
                    while n < self.col:
                        tempMatrix.data[i][j] = tempMatrix.data[i][
                            j] + (self.data[i][n] * Mb.data[n][j])
                        n = n + 1
                    j = j + 1
                i = i + 1
        return tempMatrix

    def hardamard(self, data):
        '''[summary]

        [description]

        Arguments:
            data {[type]} -- [description]

        Returns:
            | -- [description]
        '''
        if isinstance(data, Matrix):
            tempMatrix = Matrix(data.row, data.col)
            for i in range(0, data.col):
                for j in range(data.row):
                    tempMatrix.data[i][j] = self.data[i][j] * data.data[i][j]
            return tempMatrix
        if isinstance(data, Vector):
            tempVector = []
            if len(data.data) == len(self.data):
                for i in range(0, len(data.data)):
                    for j in range(0, len(self.row)):
                        tempVector = data.data[i] * self.data[i][j]
            return tempVector

    def tr(self, Ma):
        '''[summary]

        矩阵迹运算

        Arguments:
            Ma {Matrix} -- 需要迹运算的矩阵

        Returns:
            number  -- 迹运算结果值
        '''
        if isinstance(Ma, Matrix):
            total = 0
            for i in range(0, Ma.col):
                for j in range(0, Ma.row):
                    if i == j:
                        total = total + Ma.data[i][j]
            return total

    def frobenius(self, A):
        '''[summary]

        矩阵范数

        Arguments:
            A {Matrix} -- 需要求范数的矩阵

        Returns:
            number -- 范数值
        '''
        tempNums = 0
        if isinstance(A, Matrix):
            for i in range(0, A.row):
                for j in range(0, A.col):
                    tempNums = tempNums + A.data[i][j]**2
            return tempNums**0.5

    def tran(self, A):
        """矩阵转置
            A 需要被转置的矩阵
        """
        if isinstance(A, Matrix):
            NMatrix = Matrix(A.col, A.row)
            for i in range(0, A.row):
                for j in range(0, A.col):
                    NMatrix.data[j][i] = A.data[i][j]
            return NMatrix

    def map(self):
        '''[summary]

        [description]

        Returns:
            list -- 返回数组
        '''
        tempArray = []
        for x in range(0, self.row):
            for i in range(0, self.col):
                tempArray.append(self.data[x][i])
        return tempArray

    def insertData(self, *args):
        '''[summary]

        为矩阵插入数据

        Arguments:
            *args {Number} -- 数据
        '''
        n = 0
        for i in range(0, self.row):
            for j in range(0, self.col):
                if n >= len(args):
                    if isinstance(ele,(list,tuple)):
                        for x in args[n]:
                            self.data[i][j] = x
                    self.data[i][j] = args[n]
                    n = n + 1
        return self

    def pooling(self, value):
        '''矩阵池化
        ```md
        矩阵池化是将一个大的矩阵划分成小的矩阵，然后选取小的矩阵当中的最大数形成新的矩阵
        ```
        [description]

        Arguments:
            value {int} -- 一个整型数据

        Returns:
            Matrix -- 返回池化后的矩阵
        '''
        if isinstance(value, Matrix):
            h, w = value.shape
            tempList = []
            tempMatrix = Matrix(value.row,value.col)
            for i in range(h):
                for j in range(w):
                    for n in range(0,h):
                       s=self.data[i+n][j:j+w]
                       if len(s[0]) == h:tempMatrix.insertData(s)
        elif isinstance(value,(tuple,list)):
            tempMatrix = Matrix(value)
            self.pooling(tempMatrix)
        return tempMatrix

    def flat(self):
        return self.map()
if __name__ == "__main__":
    Cover = Matrix([[1,2,3],[4,5,6],[7,8,9]])
    Cover.pooling([[1,2,3],[4,5,6]])

File: statistics/TypeScript/src/Matrix.ts
import { Vector } from "./vector";
/**
 * 实现矩阵的加减以及逆运算
 */
class Matrix{
    /**
     * 矩阵行数
     */
    row:number;
    /**
     * 矩阵列数
     */
    col:number;
    /**
     * 矩阵数据
     */
    data:number[][];
    //初始化二维矩阵
    /**
     * 初始化矩阵
     * @param {number} row [行数]
     * @param {number} col [列数]
     */
    constructor(row:number =0,col:number =0,data=[]){

        if(data.length != 0 && data[0] instanceof Array){
            this.col = data.length;
            this.row = data[0].length;
        }else{
            this.row=row;//行
            this.col=col;//列
            this.data = Array(this.row);
            for(let i=0;i<this.row;i++){
                this.data[i] = new Array(this.col)
            }
        }
    }
    /**
     * 初始化为一个单位矩阵
     */
    Indentity():Matrix{
        for(let i=0;i<this.col;i++){
            for(let j=0;j<this.row;j++){
                if(i==j){
                    this.data[i][j]=1;
                }else{
                    this.data[i][j]=0;
                }
            }
        }
        return this;
    }
    /**
     * 矩阵乘法
     * @param  {Matrix} matrix [相乘矩阵]
     * @return {Matrix}        [生成的新矩阵]
     */
    product(matrix:Matrix):Matrix|undefined{
        let tempMatrix = new Matrix(this.row,matrix.col);
        if(matrix instanceof Matrix){
        if(this.col==matrix.row){
            for(let i=0;i<this.row;i++){
                for(let j=0;j<matrix.col;j++){
                    tempMatrix.data[i][j]=0;
                    for(let k=0;k<this.col;k++){
                        tempMatrix.data[i][j] += this.data[i][k]*matrix.data[k][j];
                    }
                }
            }
            return tempMatrix;
        }
    }

    }
    /**
     * 元素对应乘积
     * @param data 矩阵或者向量
     * @returns 矩阵或者是向量
     */
    Hardamard(data:Matrix|Vector):Matrix{
        if(data instanceof Matrix){
            let tempMatrix = new Matrix(data.row,data.col);
            for(let i=0;i<data.col;i++){
                for (let j = 0; j < data.row; j++) {
                    tempMatrix[i][j] = this.data[i][j]*data.data[i][j];
                }
            }
            return (this,tempMatrix);
        }
        if(data instanceof Vector){
            let tempV = new Array(data.data.length);
            if(data.data.length==this.data.length){
                for(let i=0;i<data.data.length;i++){
                    for(let j=0;j< this.row;j++){
                        tempV[i] =  data.data[i]*this.data[i][j];
                    }
                }
            }
            return tempV;
        }
    }
    /**
     * 矩阵迹运算
     * @param  {Matrix} matrix [迹运算对象]
     * @return {number}        [返回总和]
     */
    Tr():number{
        let total:number=0;
        for(let i=0;i<this.col;i++){
            for(let j=0;j<this.row;j++){
                if(i==j){
                    total+=this.data[i][j];
                }
            }
        }
        return (this,total);
    }
    /**
     * 将矩阵降维为数组
     * @param  {Matrix}        matrix [需要降维的矩阵]
     * @return {Array<number>}        [降维后返回的数组]
     */
    downDimensionality(matrix_:Matrix):Array<number>{
        let tempAry:Array<number>=[];
        if(matrix_ instanceof Matrix){
            for(let i=0;i<matrix_.col;i++){
                for(let j=0;j<matrix_.row;j++){
                    tempAry.push(matrix_[i][j]);
                }
            }
        }
        return tempAry;
    }
    /**
     * 求矩阵范数
     * @param {Matrix} A [description]
     */
    frobenius(){
        let tempNums:number=0;
        for(let i=0;i<this.row;i++){
            for(let j=0;j<this.col;j++){
                tempNums+= this.data[i][j]**2
            }
        }
        return (this,tempNums**(1/2));
    }
    /**
     * 矩阵的转置
     * [tran description]
     * @param  {Matrix} A [需被转置的矩阵]
     * @return {Matrix}   [转置后的结果]
     */
    tran():Matrix{
        for(let i=0;i<this.row;i++){
            for(let j=0;j<this.col;j++){
                let temp:number = 0;
                temp = this.data[j][i];
                this.data[j][i] = this.data[i][j];
                this.data[i][j] = temp;
            }
        }
        return this;
    }
    /**
     * @param {void} 无输入
     * @returns {Array} 将矩阵转换成一维
     */
    flat():Array<number>{
        var tempAry:Array<number> = [];
        for(let i =0;i<this.row;i++){
            for(let j=0;j<this.col;j++){
                tempAry.push(this.data[i][j]);
            }
        }
        return tempAry;
    }
}
export {Matrix};
